packages/collaboration/src/browser/collaboration-instance.ts

```diff
@@ -0,0 +1,824 @@
+    }
+
+    protected getOpenEditors(uri?: URI): EditorWidget[] {
+        const widgets = this.shell.widgets;
+        let editors = widgets.filter(e => e instanceof EditorWidget) as EditorWidget[];
+        if (uri) {
+            const uriString = uri.toString();
+            editors = editors.filter(e => e.getResourceUri()?.toString() === uriString);
+        }
+        return editors;
+    }
+
+    protected createSelectionFromRelative(selection: types.RelativeTextSelection, model: MonacoEditorModel): Selection | undefined {
+        const start = Y.createAbsolutePositionFromRelativePosition(selection.start, this.yjs);
+        const end = Y.createAbsolutePositionFromRelativePosition(selection.end, this.yjs);
+        if (start && end) {
+            return {
+                start: model.positionAt(start.index),
+                end: model.positionAt(end.index),
+                direction: selection.direction === types.SelectionDirection.LeftToRight ? 'ltr' : 'rtl'
+            };
+        }
+        return undefined;
+    }
+
+    protected createRelativeSelection(selection: Selection, model: TextEditorDocument, ytext: Y.Text): types.RelativeTextSelection {
+        const start = Y.createRelativePositionFromTypeIndex(ytext, model.offsetAt(selection.start));
+        const end = Y.createRelativePositionFromTypeIndex(ytext, model.offsetAt(selection.end));
+        return {
+            start,
+            end,
+            direction: selection.direction === 'ltr'
+                ? types.SelectionDirection.LeftToRight
+                : types.SelectionDirection.RightToLeft
+        };
+    }
+
+    protected readonly yjsMutex = createMutex();
+
+    protected registerModelUpdate(model: MonacoEditorModel): void {
+        let updating = false;
+        const modelPath = this.utils.getProtocolPath(new URI(model.uri));
+        if (!modelPath) {
+            return;
+        }
+        const unknownModel = !this.yjs.share.has(modelPath);
+        const ytext = this.yjs.getText(modelPath);
+        const modelText = model.textEditorModel.getValue();
+        if (this.isHost && unknownModel) {
+            // If we are hosting the room, set the initial content
+            // First off, reset the shared content to be empty
+            // This has the benefit of effectively clearing the memory of the shared content across all peers
+            // This is important because the shared content accumulates changes/memory usage over time
+            this.resetYjsText(ytext, modelText);
+        } else {
+            this.options.connection.editor.open(this.host.id, modelPath);
+        }
+        // The Ytext instance is our source of truth for the model content
+        // Sometimes (especially after a lot of sequential undo/redo operations) our model content can get out of sync
+        // This resyncs the model content with the Ytext content after a delay
+        const resyncDebounce = debounce(() => {
```

Comment:
But only because of `model.textEditorModel.getValue() !== newContent`. That we need this back stop tells me we don't really have a handle on the syncing logic. But fair enough...I henceforth reserve the right to summon you by your true name whenever the contents get out of sync ;-)